internal/registry/aws/auth.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/google/go-containerregistry/pkg/authn"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/fluxcd/image-reflector-controller/internal/registry"
+)
+
+// ParseImage returns the AWS account ID and region and `true` if
+// the image repository is hosted in AWS's Elastic Container Registry,
+// otherwise empty strings and `false`.
+func ParseImage(image string) (accountId, awsEcrRegion string, ok bool) {
+	registryPartRe := regexp.MustCompile(`([0-9+]*).dkr.ecr.([^/.]*)\.(amazonaws\.com[.cn]*)/([^:]+):?(.*)`)
+	registryParts := registryPartRe.FindAllStringSubmatch(image, -1)
+	if len(registryParts) < 1 {
+		return "", "", false
+	}
+	return registryParts[0][1], registryParts[0][2], true
+}
+
+// Client is a AWS ECR client which can log into the registry and return
+// authorization information.
+type Client struct {
+	*aws.Config
+}
+
+// NewClient creates a new ECR client with default configurations.
+func NewClient() *Client {
+	return &Client{Config: aws.NewConfig()}
+}
+
+// getLoginAuth obtains authentication for ECR given the account
+// ID and region (taken from the image). This assumes that the pod has
+// IAM permissions to get an authentication token, which will usually
+// be the case if it's running in EKS, and may need additional setup
+// otherwise (visit
+// https://docs.aws.amazon.com/sdk-for-go/api/aws/session/ as a
+// starting point).
+func (c *Client) getLoginAuth(accountId, awsEcrRegion string) (authn.AuthConfig, error) {
+	// No caching of tokens is attempted; the quota for getting an
+	// auth token is high enough that getting a token every time you
+	// scan an image is viable for O(1000) images per region. See
+	// https://docs.aws.amazon.com/general/latest/gr/ecr.html.
```

Comment:
I am not sure I follow the comment. From that page the `The maximum number of GetAuthorizationToken requests that you can make per second in the current Region` is 500.

Reply:
Looks like this comment is based on flux v1 https://github.com/fluxcd/image-reflector-controller/commit/0ea1627bb8fe5302e3b56c2eb99b6c961da6e739.
It's possible that the numbers have changed now and we can change it to 500.